rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,469 @@
+the resource and act appropriately.
+
+```protobuf
+import "teleport/header/v1/resourceheader.proto";
+
+// Foo is a resource that does foo.
+message Foo {
+  // The kind of resource represented.
+  string kind = 1;
+  // An optional subkind to differentiate variations of the same kind.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
+  // Common metadata that all resources shared.
+  teleport.header.v1.Metadata metadata = 4;
+  // The resource specific specification.
+  FooSpec spec = 5;
+}
+
+// FooSpec contains resource specific properties.
+message FooSpec {
+  string bar = 1;
+  int32 baz = 2;
+  bool qux = 3;
+}
+```
+
+This differs from existing resources because legacy resources make heavy use of features provided
+by [gogoprotobuf](https://github.com/gogo/protobuf). Since that project has long been abandoned, we're striving to
+migrate away from it as described in [RFD-0139](https://github.com/gravitational/teleport/pull/28386).
+The `teleport.header.v1.Metadata` is a clone of `types.Metadata` which doesn't use any of the gogoproto features.
+Legacy resources also had a `types.ResourceHeader` that used gogo magic to embed the type in the resource message. To
+get around this, the required fields from the header MUST be included in the message itself. A non-gogo clone does exist
+`teleport.header.v1.ResourceHeader`, however, to get the fields embeded custom marshalling must be manually written.
+
+### API
+
+All APIs should follow the following conventions that are largely based on
+the [Google API style guide](https://cloud.google.com/apis/design/standard_methods).
+
+#### Create
+
+The `Create` RPC takes a resource to be created and must also return the newly created resource so that any fields that
+are populated server side are provided to clients without requiring an additional call to `Get`.
+
+The request MUST fail and return a `trace.AlreadyExists` error if a matching resource is already present in the backend.
+
+```protobuf
+// Creates a new Foo resource in the backend.
+    rpc CreateFoo(CreateFooRequest) returns (CreateFooResponse);
+
+message CreateFooRequest {
+  // The desired Foo to be created.
+  Foo foo = 1;
+}
+
+message CreateFooResponse {
+  // The created Foo as it exists in the backend. The only differences between the returned
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+When creating a new resource, the `backend.Backend.Create` method should be used to persist the resource. It is also
+imperative that the revision generated by the backend is set on the returned resource.
+
+```go
+func (s *FooService) CreateFoo(ctx context.Context, foo foov1.Foo) (foov1.Foo, error) {
+value, err := convertFooToValue(foo)
```

Comment:
IMO we shouldn't be recommended these custom marshal/unmarshal functions.  Tying behavior to encoding is bad (and makes it a nightmare to handle any resource in a generic manner).

I think that we should recommend using marshal/unmarshal helpers (e.g. `utils.FastMarshal`) for encoding/decoding, and methods (e.g. `CheckAndSetDefaults`) for behavior/validation.  Case in point, all the backend-related example code in this RFD could be written as a generic helper over `T Resource` if we just got rid of the `convert*` functions.